community/kernel/src/main/java/org/neo4j/kernel/impl/index/labelscan/NativeLabelScanStore.java

```diff
@@ -0,0 +1,265 @@
+ * need to be handled externally.
+ * <p>
+ * About the {@link Layout} used in this instance of {@link GBPTree}:
+ * <ul>
+ * <li>
+ * Each keys is a combination of {@code labelId} and {@code nodeIdRange} ({@code nodeId/64}).
+ * </li>
+ * <li>
+ * Each value is a 64-bit bit set (a primitive {@code long}) where each set bit in it represents
+ * a node with that label, such that {@code nodeId = nodeIdRange+bitOffset}. Range size (e.g. 64 bits)
+ * is configurable on initial creation of the store, 8, 16, 32 or 64.
+ * </li>
+ * </ul>
+ * <p>
+ * {@link #force(IOLimiter)} is vital for allowing this store to be recoverable, and must be called
+ * whenever Neo4j performs a checkpoint.
+ * <p>
+ * This store is backed by a single store file "neostore.labelscanstore.db".
+ */
+public class NativeLabelScanStore implements LabelScanStore
+{
+    /**
+     * {@link PageCache} to {@link PageCache#map(File, int, java.nio.file.OpenOption...)}
+     * store file backing this label scan store. Passed to {@link GBPTree}.
+     */
+    private final PageCache pageCache;
+
+    /**
+     * Store file {@link PageCache#map(File, int, java.nio.file.OpenOption...)}.
+     */
+    private final File storeFile;
+
+    /**
+     * Size of each node id range, e.g. 8, 16, 32 or 64. This value is written at creation,
+     * otherwise verified against index meta data on open.
+     */
+    private final int rangeSize;
+
+    /**
+     * Used in {@link #start()} if the store is empty, where this will provide all data for fully populating
+     * this label scan store. This can be the case when changing label scan store provider on an existing database.
+     */
+    private final FullStoreChangeStream fullStoreChangeStream;
+
+    /**
+     * Page size to use for each tree node in {@link GBPTree}. Passed to {@link GBPTree}.
+     */
+    private final int pageSize;
+
+    /**
+     * The index which backs this label scan store. Instantiated in {@link #init()} and considered
+     * started after call to {@link #start()}.
+     */
+    private GBPTree<LabelScanKey,LabelScanValue> index;
+
+    /**
+     * Whether or not {@link #start()} has been called.
+     */
+    private boolean started;
+
+    /**
+     * If {@link #index} is {@code null} and {@link #started} is {@code false},
+     * then it's between {@link #init()} and {@link #start()}. If {@link #newWriter()} is called at this
+     * point we infer that recovery is taking place and so we notify the {@link GBPTree} about that fact.
+     */
+    private boolean recoveryStarted;
+
+    public NativeLabelScanStore( PageCache pageCache, File storeDir, int rangeSize, int pageSize,
+            FullStoreChangeStream fullStoreChangeStream )
+    {
+        this.pageCache = pageCache;
+        this.pageSize = pageSize;
+        this.fullStoreChangeStream = fullStoreChangeStream;
+        this.storeFile = new File( storeDir, DEFAULT_NAME + ".labelscanstore.db" );
```

Comment:
hm it looks like a new `StoreFile` type should be introduced instead and maybe name in store factory? WDYT?